File: internal/proxy/enforcer/policy.py
"""Pure egress matcher: a faithful Python port of internal/policy (the Go matcher).

Given an in-memory rule set and a request (host, path, method), ``decide`` returns
the cage's verdict -- allow, soft-deny, or hard-deny -- plus the carried enforcement
mode and the rule the decision is attributed to. It is deliberately pure: no
filesystem in the matcher itself (``load_policy`` is the one disk touch, kept
separate), no mitmproxy import. That keeps the C1 decision-vector corpus and the
golden tests runnable without mitmproxy installed.

This logic exists twice -- here in Python (for the mitmproxy enforcer, AC-0017) and
in Go (for ``policy explain``, internal/policy). They must never disagree. The
guardrail is the language-neutral corpus under internal/policy/testdata/
decision-vectors/ (cross-cutting C1), which both implementations replay. No change
to this matcher is complete without the corpus still passing.

Matching semantics (kept byte-for-byte with internal/policy/match.go + glob.go):

  - Host: case-insensitive. "*" matches any host; "*.suffix" matches any host with
    at least one label before ".suffix" (the bare apex is NOT matched); else exact.
  - Path: prefix-by-default. Pattern and path are trimmed of leading/trailing "/"
    and split into segments; the pattern must match a *prefix* of the path's
    segments. Within a segment "*" matches any run of characters; "?" and
    everything else are literal. A whole-segment "**" matches zero or more segments
    (the only token that crosses "/"); "**" glued to other characters degrades to
    "*". A rule with no paths (None) matches any path; a rule with paths matches if
    any one pattern matches.
  - Method: a rule with no methods (None) matches any method; otherwise the request
    method must be a verbatim member of the list.
  - Most-specific-wins: ordered, deterministic tie-break, independent of rule order.
  - Passthrough blind spot: a path-scoped deny_always is suppressed when the
    request's most-specific matching allow is mode "passthrough"; a host-level
    deny_always (no paths) still hard-denies a passthrough host.

IMPORTANT -- None vs empty list: an absent ``paths``/``methods`` key (or JSON null)
decodes to ``None`` and means "any" (host-wide / any method), exactly like Go's nil
slice. A present empty list ``[]`` is NOT the same: it matches no path / no method.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Optional

# Decision constants -- exhaustive, matching internal/policy/policy.go.
DECISION_ALLOW = "allow"
DECISION_SOFT_DENY = "soft-deny"
DECISION_HARD_DENY = "hard-deny"

# Enforcement modes a rule may carry.
MODE_INTERCEPT = "intercept"
MODE_PASSTHROUGH = "passthrough"

# policy.json schema version -- the cross-language contract with the Go compiler.
COMPILED_VERSION = 1


@dataclass(frozen=True)
class Rule:
    """One egress allow/deny entry as the matcher sees it.

    ``paths``/``methods`` are None when the key is omitted ("any"), mirroring Go's
    nil slice. ``source``/``lower_trust`` are stamped by the compiler for the
    on-disk artifact and are ignored by ``decide`` (absent from the corpus).
    """

    host: str
    paths: Optional[list[str]] = None
    methods: Optional[list[str]] = None
    mode: str = ""
    reason: str = ""

    @staticmethod
    def from_dict(d: dict) -> "Rule":
        return Rule(
            host=d["host"],
            paths=d.get("paths"),  # None when absent or JSON null
            methods=d.get("methods"),
            mode=d.get("mode", "") or "",
            reason=d.get("reason", "") or "",
        )


@dataclass(frozen=True)
class RuleSet:
    """The compiled, in-memory policy: the unioned allow and deny_always lists."""

    allow: list[Rule] = field(default_factory=list)
    deny_always: list[Rule] = field(default_factory=list)

    @staticmethod
    def from_dict(d: dict) -> "RuleSet":
        return RuleSet(
            allow=[Rule.from_dict(r) for r in (d.get("allow") or [])],
            deny_always=[Rule.from_dict(r) for r in (d.get("deny_always") or [])],
        )


@dataclass(frozen=True)
class Request:
    """The single egress attempt being decided."""

    host: str
    path: str
    method: str

    @staticmethod
    def from_dict(d: dict) -> "Request":
        return Request(host=d["host"], path=d["path"], method=d["method"])


@dataclass(frozen=True)
class MatchedRule:
    """Which list a decision is attributed to, and the rule's index within it.

    None for a soft-deny. The list/index form is language-neutral -- it survives two
    rules sharing a host, where a host string alone would be ambiguous.
    """

    list: str  # "allow" or "deny_always"
    index: int

    def to_dict(self) -> dict:
        return {"list": self.list, "index": self.index}


@dataclass(frozen=True)
class Result:
    """The matcher's verdict: decision, carried mode ("" for soft-deny), matched."""

    decision: str
    mode: str = ""
    matched: Optional[MatchedRule] = None


# --- the decision algorithm (internal/policy/match.go) -------------------------


def decide(rs: RuleSet, req: Request) -> Result:
    """Evaluate ``req`` against ``rs`` and return the cage's verdict.

    1. Find the most-specific matching allow; the host is "passthrough" iff that
       allow carries mode passthrough.
    2. Evaluate deny_always. A host-level deny (no paths) hard-denies even a
       passthrough host; a path-scoped deny hard-denies only on a non-passthrough
       host. The most specific *eligible* matching deny is reported.
    3. Otherwise an allow match is an allow (carrying its mode); no match is the
       implicit soft-deny.
    """
    allow_idx, allow_ok = _best_match(rs.allow, req, None)
    is_passthrough = allow_ok and rs.allow[allow_idx].mode == MODE_PASSTHROUGH

    # deny_always shadows allow. On a passthrough host only host-level denies are
    # eligible, because a path-scoped deny cannot be enforced without the path.
    eligible = None
    if is_passthrough:
        eligible = lambda r: r.paths is None  # noqa: E731

    deny_idx, deny_ok = _best_match(rs.deny_always, req, eligible)
    if deny_ok:
        r = rs.deny_always[deny_idx]
        return Result(
            decision=DECISION_HARD_DENY,
            mode=r.mode,
            matched=MatchedRule("deny_always", deny_idx),
        )

    if allow_ok:
        return Result(
            decision=DECISION_ALLOW,
            mode=rs.allow[allow_idx].mode,
            matched=MatchedRule("allow", allow_idx),
        )
    return Result(decision=DECISION_SOFT_DENY)


def _best_match(rules, req, eligible):
    """Index of the most-specific rule that matches req and passes ``eligible``.

    ``eligible`` of None means all rules are eligible. Returns (index, found). On an
    exact specificity tie the earlier index is kept (strict more-specific only).
    """
    best = -1
    best_spec = None
    for i, r in enumerate(rules):
        if eligible is not None and not eligible(r):
            continue
        pattern, ok = _rule_matches(r, req)
        if not ok:
            continue
        spec = _rule_specificity(r, pattern)
        if best == -1 or _more_specific(spec, best_spec):
            best, best_spec = i, spec
    return best, best != -1


def _rule_matches(r: Rule, req: Request):
    """Whether r matches req on host, method, and path.

    Returns (matched_pattern, ok); the matched pattern is the most specific one when
    the rule lists several, "" when the rule is host-wide.
    """
    if not _match_host(r.host, req.host):
        return "", False
    if not _match_method(r.methods, req.method):
        return "", False
    if r.paths is None:
        return "", True  # host-wide: matches any path

    matched = ""
    found = False
    for p in r.paths:
        if not _match_path(p, req.path):
            continue
        if not found or _more_specific(
            _rule_specificity(r, p), _rule_specificity(r, matched)
        ):
            matched, found = p, True
    return matched, found


# --- host / path / method matching (internal/policy/glob.go) -------------------


def _match_host(pattern: str, host: str) -> bool:
    pattern = pattern.lower()
    host = host.lower()
    if pattern == "*":
        return True
    if pattern.startswith("*."):
        # suffix includes the leading dot, e.g. ".medium.com"; endswith then
        # requires at least one label before it, so the apex does not match.
        return host.endswith(pattern[1:])
    return pattern == host


def _match_path(pattern: str, path: str) -> bool:
    return _match_segments(_split_path(pattern), _split_path(path))


def _split_path(p: str) -> list[str]:
    """Trim leading/trailing "/" then split on "/". Root ("/") yields no segments."""
    p = p.strip("/")
    if p == "":
        return []
    return p.split("/")


def _match_segments(pat: list[str], path: list[str]) -> bool:
    """Whether pat matches a prefix of path. "**" consumes zero or more segments."""
    if len(pat) == 0:
        return True
    if pat[0] == "**":
        # Consume zero segments, or consume one and retry the same "**".
        if _match_segments(pat[1:], path):
            return True
        return len(path) > 0 and _match_segments(pat, path[1:])
    if len(path) == 0:
        return False
    return _match_segment_glob(pat[0], path[0]) and _match_segments(pat[1:], path[1:])


def _match_segment_glob(pat: str, seg: str) -> bool:
    """Single-segment glob. "*" matches any run; "?" and all else are literal.

    Classic two-pointer wildcard match with backtracking on the last "*". A segment
    that is not exactly "**" but contains "**" treats each "*" independently, so
    glued "**" degrades to "*".
    """
    pi = si = 0
    star = -1
    star_match_end = 0
    while si < len(seg):
        if pi < len(pat) and pat[pi] == "*":
            star = pi
            star_match_end = si
            pi += 1
        elif pi < len(pat) and pat[pi] == seg[si]:
            pi += 1
            si += 1
        elif star >= 0:
            pi = star + 1
            star_match_end += 1
            si = star_match_end
        else:
            return False
    while pi < len(pat) and pat[pi] == "*":
        pi += 1
    return pi == len(pat)


def _match_method(methods: Optional[list[str]], method: str) -> bool:
    """None matches any method; otherwise verbatim membership ([] matches none)."""
    if methods is None:
        return True
    return method in methods


# --- specificity / tie-breaking (internal/policy/glob.go) ----------------------


@dataclass(frozen=True)
class _Specificity:
    host_rank: int  # exact 2, *.suffix 1, * 0
    path_scoped: int  # 1 if the rule constrains paths
    literal_segs: int  # count of non-wildcard segments in the matched pattern
    total_segs: int  # segment count of the matched pattern
    method_scoped: int  # 1 if the rule constrains methods
    canonical: str  # deterministic fallback for exact ties


def _rule_specificity(r: Rule, matched_pattern: str) -> _Specificity:
    canonical = (
        r.host
        + "|"
        + ",".join(r.paths or [])
        + "|"
        + ",".join(r.methods or [])
        + "|"
        + r.mode
    )
    path_scoped = 0
    literal_segs = 0
    total_segs = 0
    if r.paths is not None:
        path_scoped = 1
        segs = _split_path(matched_pattern)
        total_segs = len(segs)
        for seg in segs:
            if "*" not in seg:
                literal_segs += 1
    method_scoped = 1 if r.methods is not None else 0
    return _Specificity(
        host_rank=_host_rank(r.host),
        path_scoped=path_scoped,
        literal_segs=literal_segs,
        total_segs=total_segs,
        method_scoped=method_scoped,
        canonical=canonical,
    )


def _host_rank(host: str) -> int:
    if host == "*":
        return 0
    if host.startswith("*."):
        return 1
    return 2


def _more_specific(a: _Specificity, b: _Specificity) -> bool:
    """Whether a is strictly more specific than b (canonical: smaller wins)."""
    if a.host_rank != b.host_rank:
        return a.host_rank > b.host_rank
    if a.path_scoped != b.path_scoped:
        return a.path_scoped > b.path_scoped
    if a.literal_segs != b.literal_segs:
        return a.literal_segs > b.literal_segs
    if a.total_segs != b.total_segs:
        return a.total_segs > b.total_segs
    if a.method_scoped != b.method_scoped:
        return a.method_scoped > b.method_scoped
    return a.canonical < b.canonical


# --- connect-stage disposition (host only; no path/method visible) -------------


@dataclass(frozen=True)
class HostDisposition:
    """What the proxy can decide about a host at CONNECT / TLS-ClientHello time.

    At that stage only the host (CONNECT authority / SNI) is known -- there is no
    path or method yet -- so this is a deliberately host-granular view, consistent
    with "passthrough is host-granularity only by construction" (docs/design.md).

    - ``passthrough``: tunnel without terminating TLS. True iff the most
      host-specific matching allow is passthrough AND no equally-host-specific allow
      asks to intercept (an intercept rule forces termination so path/method rules
      can be applied; an ambiguous/mixed host resolves to intercept).
    - ``deny_reason``: set when a host-level deny_always (no paths) matches. Such a
      deny is still enforced on a passthrough host -- the tunnel is refused at
      CONNECT, since it cannot be enforced once tunnelled. (Path-scoped denies are
      invisible at this stage and handled later, per-request, on intercept hosts.)
    """

    passthrough: bool
    deny_reason: Optional[str]


def host_disposition(rs: RuleSet, host: str) -> HostDisposition:
    # Host-level deny (paths is None): the most host-specific matching one wins.
    deny_reason = None
    best_deny_rank = -1
    for r in rs.deny_always:
        if r.paths is None and _match_host(r.host, host):
            rank = _host_rank(r.host)
            if rank > best_deny_rank:
                best_deny_rank, deny_reason = rank, r.reason

    # Passthrough iff the top host-rank tier of matching allows is all passthrough.
    matching = [r for r in rs.allow if _match_host(r.host, host)]
    passthrough = False
    if matching:
        top_rank = max(_host_rank(r.host) for r in matching)
        top = [r for r in matching if _host_rank(r.host) == top_rank]
        passthrough = all(r.mode == MODE_PASSTHROUGH for r in top)

    return HostDisposition(passthrough=passthrough, deny_reason=deny_reason)


# --- policy.json loading -------------------------------------------------------


def load_policy(path: str) -> RuleSet:
    """Load a compiled policy.json ({version, input_hash, allow, deny_always}).

    ``input_hash`` is ignored (it keys the compiler's cache). Missing allow/
    deny_always lists tolerate as empty.
    """
    with open(path, "rb") as f:
        data = json.load(f)
    return RuleSet.from_dict(data)

File: internal/policy/policy.go
// Package policy is the egress matcher: given an in-memory rule set and a request
// (host, path, method), it returns the cage's decision — allow, soft-deny, or
// hard-deny — plus the carried enforcement mode and the rule the decision is
// attributed to. It is deliberately pure: no filesystem, no clock, no OS. Reading
// rules off a compiled policy.json on disk is the compiler's job (AC-0013); this
// package matches an already-in-memory RuleSet.
//
// The decision model is fixed by docs/design.md ("Network refusal handling") and is
// exhaustive: a request that matches a deny_always rule is a hard-deny; otherwise a
// request that matches an allow rule is an allow (carrying that rule's mode);
// otherwise it is a soft-deny. There is no separate "default" — soft-deny *is* the
// default. deny_always shadows allow.
//
// This logic exists twice — here in Go (for `policy explain`) and in Python (for the
// mitmproxy enforcer, AC-0017). They must never disagree. The guardrail is the
// language-neutral corpus under testdata/decision-vectors/ (cross-cutting C1): plain
// JSON `(ruleset, request) -> expected {decision, mode, matched_rule}` vectors that
// both implementations run. No change to this matcher is complete without adding or
// updating vectors.
//
// Matching semantics (the contract the Python side must reproduce exactly):
//
//   - Host: the request host is canonicalized once at the matcher entry (Decide and
//     HostDisposition) — lowercased, a trailing ":port" stripped, a single trailing "."
//     stripped — so api.example.com, API.EXAMPLE.COM, api.example.com. and
//     api.example.com:443 decide identically and a host-level deny_always cannot be
//     evaded by spelling (the Python enforcer canonicalizes identically; the corpus
//     proves it). Matching is then case-insensitive: "*" matches any host; "*.suffix"
//     matches any host with at least one label before ".suffix" (the bare apex is NOT
//     matched); else exact equality. Rule patterns are validated at config load, not
//     canonicalized here.
//   - Path: prefix-by-default. Pattern and request path are trimmed of leading and
//     trailing "/" and split into segments. A pattern matches when its segments match
//     a *prefix* of the request's segments (remaining request segments are "under"
//     the prefix and still match). Within a segment, "*" matches any run of
//     characters; "?" and everything else are literal. A whole-segment "**" matches
//     zero or more segments (the only token that crosses "/"); "**" glued to other
//     characters degrades to "*". A rule with no paths (nil) matches any path; a
//     rule with paths matches if any one pattern matches.
//   - Method: a rule with no methods (nil) matches any method; otherwise the request
//     method must be a verbatim member of the list.
//   - Most-specific-wins: when several rules in the same list match, the reported
//     rule and carried mode come from the most specific one (host exact > *.suffix >
//     *; then path constrained > host-wide, more literal segments, more segments;
//     then method constrained > any), with a deterministic fallback for exact ties.
//     This is order-independent — allow/deny_always lists union across include: files.
//   - Passthrough blind spot: a path-scoped deny_always is *suppressed* when the
//     request's most-specific matching allow is mode: passthrough, because the proxy
//     never sees the path on a passthrough host (docs/design.md). A host-level
//     deny_always (no paths) still hard-denies a passthrough host.
package policy

import "github.com/tobyS/agent-creance/internal/config"

// Decision is the cage's verdict on a request. The three values are exhaustive.
const (
	DecisionAllow    = "allow"
	DecisionSoftDeny = "soft-deny"
	DecisionHardDeny = "hard-deny"
)

// Enforcement modes a rule may carry. These mirror config.ModeIntercept /
// config.ModePassthrough but are duplicated as plain strings so the corpus (and any
// Python reader) need not know about the config package.
const (
	ModeIntercept   = "intercept"
	ModePassthrough = "passthrough"
)

// Rule is one egress allow/deny entry as the matcher sees it. Unlike config.Rule it
// uses plain []string (nil == "key omitted") and json tags, so it decodes straight
// from the language-neutral decision-vector corpus and carries no yaml/`*[]string`
// baggage from the loader.
//
// Source and LowerTrust are populated by the compiler (AC-0013) for the on-disk
// policy.json and are *ignored by Decide* — they exist so `policy show` can render an
// annotated artifact. They are absent from the decision-vector corpus (omitempty).
//
// Inject and InCage are the auth axis (AC-0068b): Inject names a credentials: entry
// the proxy will resolve and inject (AC-0068c); InCage marks a host whose auth headers
// the proxy must not touch. Like Source/LowerTrust they are annotations *ignored by
// Decide* (they act on an already-allowed request, not on the allow/deny decision), so
// they never enter the decision-vector corpus and cannot change matcher parity.
type Rule struct {
	Host       string   `json:"host"`
	Paths      []string `json:"paths,omitempty"`
	Methods    []string `json:"methods,omitempty"`
	Mode       string   `json:"mode,omitempty"`
	Reason     string   `json:"reason,omitempty"`
	Inject     string   `json:"inject,omitempty"`
	InCage     bool     `json:"in_cage,omitempty"`
	Source     string   `json:"source,omitempty"`
	LowerTrust bool     `json:"lower_trust,omitempty"`
}

// RuleSet is the compiled, in-memory policy the matcher evaluates: the unioned
// soft-allow and hard-deny lists.
type RuleSet struct {
	Allow      []Rule `json:"allow,omitempty"`
	DenyAlways []Rule `json:"deny_always,omitempty"`
}

// CompiledVersion is the policy.json schema version — the cross-language contract with
// the Python enforcer. Bump only on a breaking change to the artifact shape.
const CompiledVersion = 1

// Credential is one compiled credentials: entry (AC-0068b) as the proxy will read it:
// a reference the enforcer resolves host-side and injects, never a resolved value. It
// mirrors config.Credential but carries json tags for the artifact. Source is an
// op:// / keychain:// / env:// reference; Header is the target header; Template is the
// value-template; Username is the Basic sentinel. GitHubApp/OAuth2 are the minted
// forms (AC-0069a) — still reference-only: Key/RefreshToken are secret *references*,
// the rest is non-secret minting config. No secret value ever appears here.
//
// The Python enforcer ignores the minting blocks entirely: the token it injects
// comes from the broker by credential name, resolved (static) or minted (AC-0069a)
// host-side, and it renders through Header/Template/Username regardless of form.
type Credential struct {
	Source    string         `json:"source,omitempty"`
	Header    string         `json:"header,omitempty"`
	Template  string         `json:"template"`
	Username  string         `json:"username,omitempty"`
	GitHubApp *GitHubAppMint `json:"github_app,omitempty"`
	OAuth2    *OAuth2Mint    `json:"oauth2,omitempty"`
}

// GitHubAppMint is the compiled github_app: block (reference-only): Key is a secret
// reference to the PKCS#1 PEM app private key, the rest is non-secret config.
type GitHubAppMint struct {
	Key         string            `json:"key"`
	ClientID    string            `json:"client_id"`
	Repo        string            `json:"repo"`
	Permissions map[string]string `json:"permissions,omitempty"`
}

// OAuth2Mint is the compiled oauth2: block (reference-only): RefreshToken is a secret
// reference, the rest is non-secret config.
type OAuth2Mint struct {
	RefreshToken  string   `json:"refresh_token"`
	ClientID      string   `json:"client_id"`
	TokenEndpoint string   `json:"token_endpoint,omitempty"`
	Scopes        []string `json:"scopes,omitempty"`
}

// CredentialsFromConfig converts the config credentials map into the compiled form
// (reference-only, no resolution). It returns nil for an empty input so the artifact
// omits the block entirely.
func CredentialsFromConfig(in map[string]config.Credential) map[string]Credential {
	if len(in) == 0 {
		return nil
	}
	out := make(map[string]Credential, len(in))
	for name, c := range in {
		cc := Credential{
			Source:   c.Source,
			Header:   c.Header,
			Template: c.Template,
			Username: c.Username,
		}
		if c.GitHubApp != nil {
			cc.GitHubApp = &GitHubAppMint{
				Key:         c.GitHubApp.Key,
				ClientID:    c.GitHubApp.ClientID,
				Repo:        c.GitHubApp.Repo,
				Permissions: c.GitHubApp.Permissions,
			}
		}
		if c.OAuth2 != nil {
			cc.OAuth2 = &OAuth2Mint{
				RefreshToken:  c.OAuth2.RefreshToken,
				ClientID:      c.OAuth2.ClientID,
				TokenEndpoint: c.OAuth2.TokenEndpoint,
				Scopes:        c.OAuth2.Scopes,
			}
		}
		out[name] = cc
	}
	return out
}

// Compiled is the on-disk policy.json the compiler (AC-0013) writes and the proxy
// enforcer reads: a versioned, input-hash-keyed RuleSet whose rules carry source
// annotations. The embedded RuleSet promotes allow/deny_always to the top level, so the
// artifact serializes as {version, input_hash, allow, deny_always}, plus the optional
// credentials: block (AC-0068b, references only). InputHash keys the compiler's
// regeneration cache; the enforcer ignores it.
type Compiled struct {
	Version     int                   `json:"version"`
	InputHash   string                `json:"input_hash"`
	Credentials map[string]Credential `json:"credentials,omitempty"`
	RuleSet
}

// Request is the single egress attempt being decided.
type Request struct {
	Host   string `json:"host"`
	Path   string `json:"path"`
	Method string `json:"method"`
}

// MatchedRule identifies the rule a decision is attributed to: which list, and its
// index within that list. It is nil for a soft-deny (nothing matched). The list/index
// form is language-neutral — it survives two rules sharing a host, where a host
// string alone would be ambiguous.
type MatchedRule struct {
	List  string `json:"list"` // "allow" or "deny_always"
	Index int    `json:"index"`
}

// Result is the matcher's verdict: the decision, the carried enforcement mode (the
// matched rule's mode; "" for a soft-deny), and the matched rule (nil for a
// soft-deny).
type Result struct {
	Decision string       `json:"decision"`
	Mode     string       `json:"mode"`
	Matched  *MatchedRule `json:"matched_rule"`
}

// HostDisposition is what the proxy can decide about a host at CONNECT / TLS
// ClientHello time, when only the host is known (no path or method). It mirrors the
// Python enforcer's host_disposition (internal/proxy/enforcer/policy.py) and is driven
// by the same decision-vector corpus (the host_disposition expectation), so the two
// CONNECT-stage implementations are provably consistent (AC-0058 / C3).
//
//   - Passthrough: tunnel without terminating TLS. True iff the top host-rank tier of
//     matching allows is entirely passthrough (any intercept in that tier forces TLS
//     termination so per-request path/method rules can apply — a mixed host resolves to
//     intercept).
//   - DenyReason: set ("" when none) when a host-level deny_always (no paths) matches;
//     the most host-specific one wins. Such a deny is enforced at CONNECT even on a
//     passthrough host (the tunnel is refused), since it cannot be enforced once tunnelled.
type HostDisposition struct {
	Passthrough bool   `json:"passthrough"`
	DenyReason  string `json:"deny_reason"`
}

// HostDisposition decides the CONNECT-stage disposition of host. It canonicalizes the
// host at entry exactly as Decide does, then applies the host-only projection of the
// decision model. The Decide rule (most-specific-allow) is the canonical one; this is
// its host-granular approximation for the stage where no path/method is visible yet.
func (rs RuleSet) HostDisposition(host string) HostDisposition {
	host = canonicalHost(host)

	denyReason := ""
	bestDenyRank := -1
	for _, r := range rs.DenyAlways {
		if r.Paths == nil && matchHost(r.Host, host) {
			if rank := hostRank(r.Host); rank > bestDenyRank {
				bestDenyRank, denyReason = rank, r.Reason
			}
		}
	}

	topRank := -1
	for _, r := range rs.Allow {
		if matchHost(r.Host, host) {
			if hr := hostRank(r.Host); hr > topRank {
				topRank = hr
			}
		}
	}
	passthrough := topRank >= 0
	if passthrough {
		for _, r := range rs.Allow {
			if matchHost(r.Host, host) && hostRank(r.Host) == topRank && r.Mode != ModePassthrough {
				passthrough = false
				break
			}
		}
	}

	return HostDisposition{Passthrough: passthrough, DenyReason: denyReason}
}

// FromConfig converts a validated config.Egress into a matcher RuleSet, dereferencing
// the loader's *[]string Paths/Methods (a nil pointer stays a nil slice — "key
// omitted"). It is the bridge a later `policy explain` / compiler (AC-0013) uses to
// build a RuleSet from parsed config.
func FromConfig(e config.Egress) RuleSet {
	return RuleSet{
		Allow:      rulesFromConfig(e.Allow),
		DenyAlways: rulesFromConfig(e.DenyAlways),
	}
}

func rulesFromConfig(in []config.Rule) []Rule {
	if len(in) == 0 {
		return nil
	}
	out := make([]Rule, len(in))
	for i, r := range in {
		out[i] = RuleFromConfig(r)
	}
	return out
}

// RuleFromConfig converts one config.Rule into a matcher Rule, dereferencing the
// loader's *[]string Paths/Methods (a nil pointer stays a nil slice — "key omitted").
// Source is left empty for the caller (the compiler) to stamp with provenance.
func RuleFromConfig(r config.Rule) Rule {
	out := Rule{
		Host:   r.Host,
		Mode:   r.Mode,
		Reason: r.Reason,
		Inject: r.Inject,
		InCage: r.InCage,
	}
	if r.Paths != nil {
		out.Paths = *r.Paths
	}
	if r.Methods != nil {
		out.Methods = *r.Methods
	}
	return out
}
